Fix ragged array (#2842)
hyperspy/hyperspy: doc/user_guide/signal.rst

       >>> s
       <ComplexSignal1D, title: , dimensions: (20, 10|100)>

.. _signal.ragged:

Ragged array
------------

A ragged array (also called jagged array) is an array created with
sequences-of-sequences, where the nested sequences don't have the same length.
For example, a numpy ragged array can be created as follow:

   .. code-block:: python

       >>> arr = np.array([[1, 2, 3], [1]], dtype=object)
       >>> arr
       array([list([1, 2, 3]), list([1])], dtype=object)
       
Note that the array shape is (2, ):

   .. code-block:: python

       >>> arr.shape
       (2, )

Numpy ragged array must have python ``object`` type to allow the variable length of
the nested sequences - here ``[1, 2, 3]`` and ``[1]``. As explained in
`NEP-34 <https://numpy.org/neps/nep-0034-infer-dtype-is-object.html>`_,

> ``dtype=object`` needs to be specify when creating the array to avoid ambiguity

Reviewer: ```suggestion
``dtype=object`` needs to be specified when creating the array to avoid ambiguity
```

about the shape of the array.

HyperSpy supports the use of ragged array with the following conditions:

> - The signal must be explicitely define as being ``ragged``, either when creating

Reviewer: ```suggestion
- The signal must be explicitly defined as being ``ragged``, either when creating
```

>   the signal or by changing the ragged attribute of the signal.

Reviewer: ```suggestion
  the signal or by changing the ragged attribute of the signal
```

            <undefined> |      2 |      0 |       0 |       1 | <undefined> 
       ---------------- | ------ | ------ | ------- | ------- | ------ 
            Ragged axis |               Variable length

.. note::
    When possible, numpy will cast sequences-of-sequences to "non-ragged" array:

       .. code-block:: python

           >>> arr = np.array([np.array([1, 2]), np.array([1, 2])], dtype=object)
           >>> arr
           array([[1, 2],
                  [1, 2]], dtype=object)

       
    Unlike in the previous example, here the array is not ragged, because
    the length of the nested sequences are equal (2) and numpy will create
    an array of shape (2, 2) instead of (2, ) as in the previous example of
    ragged array

       .. code-block:: python

           >>> arr.shape
           (2, 2)

In addition to the use of the keyword ``ragged`` when creating an hyperspy
signal, the ``ragged`` attribute can also be set to specify whether the signal
contains a ragged array or not.

In the following example, an hyperspy signal is created without specifying that
the array is ragged. In this case, the signal dimension is 2, which *can be*

> misleading, because each item contains a list of number. To provide a clear

Reviewer: ```suggestion
misleading, because each item contains a list of numbers. To provide a clear
```
